src/librvnpal/src/posix/sync.c

```diff
@@ -10,134 +10,163 @@
+    char *link_name = NULL;
+    int32_t rc;
+
+    int32_t steps = 10;
+
+    while (1)
+    {
+        if (lstat(dir_path, &sb) == -1)
+        {
+            rc = FAIL_STAT_FILE;
+            goto error_cleanup;
+        }
+
+        link_name = malloc(sb.st_size + 1);
+        if (link_name == NULL)
+        {
+            rc = FAIL_NOMEM;
+            goto error_cleanup;
+        }
+
+        int32_t len = readlink(dir_path, link_name, sb.st_size + 1);
+
+        if (len == 0 || (len == -1 && errno == EINVAL))
+        { /* EINVAL on non-symlink dir_path */
+            rc = rvn_sync_directory_for_internal(dir_path, detailed_error_code);
+            goto success;
+        }
+
+        if (len < 0)
+        {
+            rc = FAIL_STAT_FILE;
+            goto error_cleanup;
+        }
+
+        if (len > sb.st_size)
+        {
+            /* race: the link has changed, re-read */
+            free(link_name);
+            link_name = NULL;
+            if (steps-- > 0)
+                continue;
+            rc = FAIL_RACE_RETRIES;
+            goto error_cleanup;
+        }
+
+        link_name[len] = '\0';
+        break;
+    }
+
+    if (depth == 0)
+    {
+        rc = FAIL_PATH_RECURSION;
+        goto error_cleanup;
+    }
+
+    rc = sync_directory_maybe_symblink(link_name,
+                                       depth - 1,
+                                       detailed_error_code);
 
 error_cleanup:
-  *detailed_error_code = errno;
+    *detailed_error_code = errno;
 success:
-  if (link_name != NULL)
-    free (link_name);
+    if (link_name != NULL)
+        free(link_name);
 
-  return rc;
+    return rc;
 }
 
-EXPORT int32_t
-sync_directory_for (const char *file_path, uint32_t * detailed_error_code) {
-  assert (file_path != NULL);
-
-  char *file_path_copy = NULL;
-  file_path_copy = strdup (file_path);
-  if (file_path_copy == NULL) {
-    *detailed_error_code = errno;
-    return FAIL_NOMEM;
-  }
-  char *dir_path = dirname (file_path_copy);
-
-  int32_t rc = sync_directory_maybe_symblink (dir_path, 
-    256, /* even that is probably just abuse */
-    detailed_error_code);
-
-  free (file_path_copy);
+int32_t
+rvn_sync_directory_for(const char *file_path, int32_t *detailed_error_code)
+{
+    assert(file_path != NULL);
+
+    char *file_path_copy = NULL;
+    file_path_copy = strdup(file_path);
+    if (file_path_copy == NULL)
+    {
+        *detailed_error_code = errno;
+        return FAIL_NOMEM;
+    }
+    char *dir_path = dirname(file_path_copy);
+
+    int32_t rc = sync_directory_maybe_symblink(dir_path,
+                                               256, /* even that is probably just abuse */
+                                               detailed_error_code);
+
+    free(file_path_copy);
+
+    return rc;
+}
 
-  return rc;
+int32_t
+rvn_memory_sync(void *address, int64_t size, int32_t flags, int32_t *detailed_error_code)
+{
+    int32_t msync_flags = 0;
+    if (flags & MSYNC_OPTIONS_MS_SYNC)
```

Comment:
Nope, we aren't going to expose the same interface as the OS has, we need to expose our own, much simpler interface